packages/ra-core/src/i18n/useSetLocale.tsx

```diff
@@ -30,9 +33,31 @@ import { TranslationContext } from './TranslationContext';
  *     );
  * }
  */
-const useSetLocale = () => {
-    const { setLocale } = useContext(TranslationContext);
-    return setLocale;
+const useSetLocale = (): SetLocale => {
+    const { setLocale, i18nProvider } = useContext(TranslationContext);
+    const { startLoading, stopLoading } = useUpdateLoading();
+    const notify = useNotify();
+    return useCallback(
+        (newLocale: string) =>
+            new Promise(resolve => {
+                startLoading();
+                // so we systematically return a Promise for the messages
+                // i18nProvider may return a Promise for language changes,
+                resolve(i18nProvider(I18N_CHANGE_LOCALE, newLocale));
```

Comment:
Can't we end up with a promise resolving to another promise here ?

Reply:
Promises are unwrapped by JS.